packages/shared/src/permissions/permissionsClass.ts

```diff
@@ -733,6 +791,48 @@ export class Permissions {
     );
   }
 
+  private checkEnvFilterUpdatePermission(
+    existing: { projects?: string[]; environment?: string },
+    updates: { projects?: string[]; environment?: string },
+    permission: EnvScopedPermission
+  ): boolean {
+    if (
+      !this.checkEnvFilterPermission(
+        existing,
+        existing.environment ? [existing.environment] : [],
+        permission
+      )
+    ) {
+      return false;
+    }
+
+    if ("projects" in updates) {
+      if (
+        !this.checkEnvFilterPermission(
+          updates,
+          existing.environment ? [existing.environment] : [],
+          permission
+        )
+      ) {
+        return false;
+      }
+    }
+
+    if ("environment" in updates) {
+      if (
+        !this.checkEnvFilterPermission(
+          existing,
+          updates.environment ? [updates.environment] : [],
+          permission
+        )
+      ) {
+        return false;
+      }
+    }
```

Comment:
I think this is missing the combo case when both environment and projects are changing.  We're not checking new environment / new projects.  Also, you could get into a weird situation where you have permission for both `existing` and the end result, but not one of these in-between states.  For example, going from `Project A / dev` to `Project B / prod`.  We shouldn't care whether or not you have permission for `Project A / prod` or `Project B / dev`.

Instead of 4 if blocks, I think we can simplify to just having 2.  First test permissions on `existing`.  Then, check permissions on `{...existing, ...updates}`.

Reply:
Just updated!